planning/behavior_velocity_planner/src/scene_module/speed_bump/scene.cpp

```diff
@@ -0,0 +1,195 @@
+}
+
+bool SpeedBumpModule::modifyPathVelocity(
+  PathWithLaneId * path, [[maybe_unused]] StopReason * stop_reason)
+{
+  debug_data_ = DebugData();
+  debug_data_.base_link2front = planner_data_->vehicle_info_.max_longitudinal_offset_m;
+
+  const auto & ego_path = *path;
+
+  path_intersects_.clear();
+
+  const auto & ego_pos = planner_data_->current_pose.pose.position;
+  const auto & speed_bump = speed_bump_reg_elem_.speedBump();
+
+  const auto & intersects =
+    getPolygonIntersects(ego_path, lanelet::utils::to2D(speed_bump).basicPolygon(), ego_pos, 2);
+
+  for (const auto & p : intersects) {
+    path_intersects_.push_back(createPoint(p.x(), p.y(), ego_pos.z));
+  }
+
+  debug_data_.path_polygon_intersection_points = path_intersects_;
+
+  for (const auto & p : speed_bump_reg_elem_.speedBump().basicPolygon()) {
+    debug_data_.speed_bump_polygon.push_back(createPoint(p.x(), p.y(), ego_pos.z));
+  }
+
+  return applySlowDownSpeed(*path);
+}
+
+bool SpeedBumpModule::applySlowDownSpeed(PathWithLaneId & output)
+{
+  if (path_intersects_.empty()) {
+    return false;
+  }
+
+  const auto & ego_pos = planner_data_->current_pose.pose.position;
+  const auto & ego_path = output;
+
+  // the range until to the point where ego will start accelerate
+  auto slow_end_point_range =
+    calcSignedArcLength(ego_path.points, ego_pos, path_intersects_.back());
+  const auto & slow_end_margin_to_base_link =
+    planner_data_->vehicle_info_.rear_overhang_m + planner_param_.slow_end_margin;
+  slow_end_point_range += slow_end_margin_to_base_link;
+
+  const auto & p_slow_end =
+    calcLongitudinalOffsetPoint(ego_path.points, ego_pos, slow_end_point_range);
+
+  if (!p_slow_end) {
+    return false;
+  }
+
+  debug_data_.slow_end_points.push_back(p_slow_end.get());
+
+  if (!passed_slow_start_point_) {
+    state_ = State::SLOW_DOWN;
+
+    // the range until to the point where ego will have a const slow down speed
+    auto slow_start_point_range =
+      calcSignedArcLength(ego_path.points, ego_pos, path_intersects_.front());
+    const auto & slow_start_margin_from_base_link =
+      planner_data_->vehicle_info_.max_longitudinal_offset_m + planner_param_.slow_start_margin;
+    slow_start_point_range -= slow_start_margin_from_base_link;
+
+    const auto & p_slow_start =
+      calcLongitudinalOffsetPoint(ego_path.points, ego_pos, slow_start_point_range);
+
+    if (!p_slow_start) {
+      return false;
+    }
+
+    insertDecelPointWithDebugInfo(p_slow_start.get(), speed_bump_slow_down_speed_, output);
+
+    if (slow_start_point_range < 0.0) {
+      passed_slow_start_point_ = true;
+    }
+  } else if (slow_end_point_range > 0) {
+    state_ = State::INSIDE;
+
+    // insert constant ego speed until the end of the speed bump
+    for (auto & p : output.points) {
+      const auto & original_velocity = p.point.longitudinal_velocity_mps;
+      p.point.longitudinal_velocity_mps = std::min(original_velocity, speed_bump_slow_down_speed_);
+    }
```

Comment:
What is INSIDE state? It would be super great if you add a state transition figure in the doc as follows.
https://autowarefoundation.github.io/autoware.universe/main/control/trajectory_follower/design/pid_longitudinal_controller-design/
![image](https://user-images.githubusercontent.com/20228327/204192446-dc76810a-5a0c-4481-bdde-15411d68276f.png)

I don't understand the intention from the line 162 to 165.

Reply:
@takayuki5168 

![speed_bump_diagram](https://user-images.githubusercontent.com/48479081/205722753-8469c99d-d01b-4819-86e7-9a7f73016739.png)

> I don't understand the intention from the line 162 to 165

Intention is to insert const speed to the path while the ego is on the speed bump

Reply:
@mehmetdogru
In the INSIDE state, now you set 0 velocity to all the path points, right?
In the same way as other modules in behavior_velocity_planner like occlusion_spot, run_out, ..., you'd better set the low velocity only to path points where there is a speed bump. Then motion_velocity_smoother will smooth the velocity.

Does this make sense to you?

Reply:
@takayuki5168 

> In the INSIDE state, now you set 0 velocity to all the path points, right?
Not 0 speed but calculated proper speed for the speed_bump (i.e 8 kph)

> Does this make sense to you?
Sure, now in `INSIDE` state [calculated slow velocity inserted on the path only until slow_end_point](https://github.com/beyzanurkaya/autoware.universe/blob/6f55fd2f38af0b19dabe2744cdd1e49f82ac6096/planning/behavior_velocity_planner/src/scene_module/speed_bump/scene.cpp#L158-L177).

Do you think that is ok?

Reply:
@takayuki5168 

On the other modules like `run_out`, isn't the implementation like below:

![sb](https://user-images.githubusercontent.com/48479081/206758052-fff6ac68-4cfc-4518-aba8-e90d6515fb07.png)

https://autowarefoundation.github.io/autoware.universe/main/planning/behavior_velocity_planner/run-out-design/#insert-velocity-to-approach-the-obstacles

Could you please elaborate on your idea? Am I missing something?

Reply:
@mitsudome-r @takayuki5168 

We decided to go without considering the first intersection point between path and the speed_bump polygon.

- We would consider only the last intersection point and calculation of `slow_end_point` and `slow_start_point` would be done wrt to this point.

wrt  `path_intersects_.back()` ->

```cpp
 slow_start_point = calcLongitudinalOffsetPoint(path, intersect_points.back(), calcSignedArcLength (speed_bump_width + start_margin))
```
So if the path is curvy then the calculation would be wrong as I tried to explain below. I guess this situation brings us back that we need to know where the first intersection point is.

![speed_bump_curvy_path_issue](https://user-images.githubusercontent.com/48479081/207884090-65226b65-4d37-429b-98b8-cc161ef156de.png)

In conclusion I will go ahead with the idea where we need to know both intersection points and try to implement simpler version of current state.

If you have any kind of opinion or objection about this please comment.

Reply:
I see.
One simple workaround that I can think of is to calculate speed_bump_width_arc_length in modifyPathVelocity.

Example of pseudo code
```
// constructor
SpeedBumpModule::SpeedBumpModule(){
    speed_bump_arc_length_ = 0.0; // default value
}

// update speed_bump_arc_length_ in modifyPathVelocity
bool SpeedBumpModule::modifyPathVelocity(){
    const auto & intersects =
    getPolygonIntersects(ego_path, lanelet::utils::to2D(speed_bump).basicPolygon(), ego_pos, 2);

    if (intersects.size() ==2){
         speed_bump_arc_length_ = getArcLength(path, intersects.front(), intersects.back());
    }

   return applySlowDownSpeed(path);
}
```

But if there are any better approach that you can think of, that is also fine with me.

Reply:
Thank you @mitsudome-r 

Yeap this is of course a solution, however you are using `intersects.front()`. My point is that without knowing where is the second intersection point there is no way of calculating the arc length of the speed bump on the path covering every case also considering the path is changing of course on every cycle.

If we are to consider both intersection point then I would rather use both intersection points separately to calculate outer points (slow_start_point & slow_end_point) instead of calculating the arc length of the speed_bump.

Reply:
@mitsudome-r @takayuki5168 

I updated the module. Removed the state of the ego, instead there is `intersection status` to classify the first and the second intersection points between path and the speed bump polygon. Please observe the path velocity profile marker that shows only in the speed bump area slow_down_speed is inserted on the related path points. 

Kindly waiting for review.

https://user-images.githubusercontent.com/48479081/208469134-1f19d4c8-3aea-40b1-b44e-d7e62345eb26.mp4

![Screenshot from 2022-12-19 11-51-45](https://user-images.githubusercontent.com/48479081/208469059-2c879ca5-79c1-4779-8bc3-6db13ae7f67e.png)

![Screenshot from 2022-12-19 11-50-56](https://user-images.githubusercontent.com/48479081/208469220-2d95bda6-0200-41aa-ae7b-f71ce97c947d.png)

Reply:
@mitsudome-r @takayuki5168 

Let me give you more information about the current status of the PR.

![SpeedBump-Design drawio](https://user-images.githubusercontent.com/48479081/208666004-3a2b85a9-8216-4a63-ac0c-846dfebc6f09.svg)

As we agreed, `slow_down_speed` is assigned only to those path points between `slow_start_point` and `slow_end_point`. To manage that I calculate  `slow_start_point` and  `slow_end_point` then assign `slow_down_speed` between these two. Here is the scenarios the module supports at the moment:

![SpeedBump-Scenarios drawio](https://user-images.githubusercontent.com/48479081/208667065-3af9ebb4-98e3-4a70-a206-fdfd00035c89.svg)

Reply:
@takayuki5168 

Is this thread resolved? If so can you please review and approve the changes if there is no problems/blockers on your end.

Reply:
Regarding this thread, looks good to me.